test_utils.go

```diff
@@ -247,6 +264,7 @@ func (a TestClient) LongCallResult(ctx context.Context) (ret int, err error) {
 
 type mockCodec struct {
 	elems []interface{}
+	mtx   sync.Mutex
```

Comment:
Bleh, I guess a lock protecting an array is the most straightforward way. Maybe put a comment explaining why it was done that way? And then make sure the lock covers all the other reads (NumElements, Decode).